### DISCUSSION TITLE
EVM Revert on FundSubscription in onTokenTranser Function in Smart contract lottery project

### DISCUSSION THREAD

**Discussion by Parth-Sharma-tonchi** [Cyfrin/foundry-full-course-cu#475] [Q&A] [answered]
EVM revert on fundSubscription.
The Error is:
`LinkToken::transferAndCall(VRFCoordinatorV2: [0x8103B0A8A00be2DDC778e6e7eaa21791Cd364625], 1000000000000000000 [1e18], 0x000000000000000000000000779877a7b0d9e8603169ddbd7836e478b4624789) 
    │   ├─ emit Transfer(from: 0x5D74A1e8F2154ab57a31C1449Dbd4BCf2a6dA693, to: VRFCoordinatorV2: [0x8103B0A8A00be2DDC778e6e7eaa21791Cd364625], value: 1000000000000000000 [1e18])
    │   ├─ emit Transfer(from: 0x5D74A1e8F2154ab57a31C1449Dbd4BCf2a6dA693, to: VRFCoordinatorV2: [0x8103B0A8A00be2DDC778e6e7eaa21791Cd364625], value: 1000000000000000000 [1e18], data: 0x000000000000000000000000779877a7b0d9e8603169ddbd7836e478b4624789)
    │   ├─ [2907] VRFCoordinatorV2::onTokenTransfer(0x5D74A1e8F2154ab57a31C1449Dbd4BCf2a6dA693, 1000000000000000000 [1e18], 0x000000000000000000000000779877a7b0d9e8603169ddbd7836e478b4624789) 
    │   │   └─ ← "EvmError: Revert"
    │   └─ ← "EvmError: Revert"
    └─ ← "EvmError: Revert"`

My FundSubscription contract is:
`contract FundSubscription is Script {
    uint96 private constant FUND_AMOUNT = 1 ether;

    function fundSubscriptionConfig() public {
        HelperConfig helperConfig = new HelperConfig();
        ( address vrfCoordinator,
        ,
        ,
        uint64 subId,
        ,
        ,
        address link) = helperConfig.currentRaffleConfig();
        return fundSubscription(vrfCoordinator, subId, link);
    }

    function fundSubscription(address vrfCoordinator, uint64 subId, address link) public {
        console.log("Funding subscription: ",subId);
        console.log("Using Coordinator :", vrfCoordinator);
        console.log("On chainId: ", block.chainid);

        if(block.chainid == 31337){
            vm.startBroadcast();
            VRFCoordinatorV2Mock(vrfCoordinator).fundSubscription(subId, FUND_AMOUNT);
            vm.stopBroadcast();
        }else{
            vm.startBroadcast();
            LinkToken(link).transferAndCall(vrfCoordinator,uint256(FUND_AMOUNT), abi.encode(link));
            vm.stopBroadcast();
        }
    }
    function run() external {
        return fundSubscriptionConfig();
    }
}`

LinkToken.sol is also here:
`// SPDX-License-Identifier: MIT

// @dev This contract has been adapted to fit with dappTools
pragma solidity ^0.8.0;

import "@solmate/tokens/ERC20.sol";

interface ERC677Receiver {
    function onTokenTransfer(
        address _sender,
        uint256 _value,
        bytes memory _data
    ) external;
}

contract LinkToken is ERC20 {
    uint256 constant INITIAL_SUPPLY = 1000000000000000000000000;
    uint8 constant DECIMALS = 18;

    constructor() ERC20("LinkToken", "LINK", DECIMALS) {
        _mint(msg.sender, INITIAL_SUPPLY);
    }

    event Transfer(
        address indexed from,
        address indexed to,
        uint256 value,
        bytes data
    );

    /**
     * @dev transfer token to a contract address with additional data if the recipient is a contact.
     * @param _to The address to transfer to.
     * @param _value The amount to be transferred.
     * @param _data The extra data to be passed to the receiving contract.
     */
    function transferAndCall(
        address _to,
        uint256 _value,
        bytes memory _data
    ) public virtual returns (bool success) {
        super.transfer(_to, _value);
        // emit Transfer(msg.sender, _to, _value, _data);
        emit Transfer(msg.sender, _to, _value, _data);
        if (isContract(_to)) {
            contractFallback(_to, _value, _data);
        }
        return true;
    }

    // PRIVATE

    function contractFallback(
        address _to,
        uint256 _value,
        bytes memory _data
    ) private {
        ERC677Receiver receiver = ERC677Receiver(_to);
        receiver.onTokenTransfer(msg.sender, _value, _data);
    }

    function isContract(address _addr) private view returns (bool hasCode) {
        uint256 length;
        assembly {
            length := extcodesize(_addr)
        }
        return length > 0;
    }
}`

Please help me to get rid of this ERROR.

**Comment by Niferu** [comment]
Hello @Parth-Sharma-tonchi

Please link your repository and tell us after running what exact command you are getting this error.

**Reply by Parth-Sharma-tonchi** [reply]
Following is the command:
`$ forge script script/interactions.s.sol:FundSubscription --rpc-url $SEPOLIA_RPC_URL --private-key $PRIVATE_KEY --broadcast`

It is an ongoing project so I didn't push it to my GitHub. But Following are the Deploy and interactions script.

Deploy Script
`

pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {Raffle} from "../src/Raffle.sol";
import {HelperConfig} from "./HelperConfig.s.sol";
import {CreateSubscription, FundSubscription, AddConsumer} from "./interactions.s.sol";

contract DeployRaffle is Script{


    function run() external returns(Raffle, HelperConfig){
        HelperConfig helperConfig = new HelperConfig();
        (address vrfCoordinatorV2,
        uint256 entranceFee,
        uint256 interval,
        uint64 subscriptionId,
        bytes32 gasLane,
        uint32 callbackGasLimit, address link) = helperConfig.currentRaffleConfig();

        if(subscriptionId == 0){
            //Creating a subscription for chainlink VRFCoordinator;
            CreateSubscription createSubscription = new CreateSubscription();
            subscriptionId = createSubscription.run();

            //Fund it
            FundSubscription fundSubscription = new FundSubscription();
            fundSubscription.fundSubscription(vrfCoordinatorV2, subscriptionId, link);
        }
        vm.startBroadcast();
        Raffle raffle = new Raffle(
            vrfCoordinatorV2, entranceFee, interval, subscriptionId, gasLane, callbackGasLimit
        );
        vm.stopBroadcast();

        AddConsumer addConsumer = new AddConsumer();
        addConsumer.addConsumer(address(raffle), vrfCoordinatorV2, subscriptionId);
        
        return (raffle, helperConfig);
    }
}`

Interactions.s.sol
`

pragma solidity ^0.8.18;

import {Script, console} from "forge-std/Script.sol";
import {HelperConfig} from "./HelperConfig.s.sol";
import {VRFCoordinatorV2Mock} from "lib/chainlink-brownie-contracts/contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol";
import {LinkToken} from "../test/mocks/LinkToken.sol";
import {DevOpsTools} from "lib/foundry-devops/src/DevOpsTools.sol";

contract CreateSubscription is Script {
    
    function run() external returns(uint64){
        return createSubscriptionAndVRFCoordinator();
    }

    function createSubscriptionAndVRFCoordinator() public returns(uint64){
        HelperConfig helperConfig = new HelperConfig();
        (address vrfCoordinator, , , , , ,) = helperConfig.currentRaffleConfig();
        return createSubscription(vrfCoordinator);
    }

    function createSubscription(address vrfCoordinator) public returns(uint64){
        console.log("Creating subscription on chainid ", block.chainid);
        console.log("Subscription Created!");
        vm.startBroadcast();
        uint64 subId = VRFCoordinatorV2Mock(vrfCoordinator).createSubscription();
        vm.stopBroadcast();
        console.log("Your subId is: ", subId);
        console.log("Please update subscriptionId in HelperConfig.s.sol");
        return subId;
    }
}

contract FundSubscription is Script {
    uint96 private constant FUND_AMOUNT = 3 ether;

    function fundSubscriptionConfig() public {
        HelperConfig helperConfig = new HelperConfig();
        ( address vrfCoordinator,
        ,
        ,
        uint64 subId,
        ,
        ,
        address link) = helperConfig.currentRaffleConfig();
        return fundSubscription(vrfCoordinator, subId, link);
    }

    function fundSubscription(address vrfCoordinator, uint64 subId, address link) public {
        console.log("Funding subscription: ",subId);
        console.log("Using Coordinator :", vrfCoordinator);
        console.log("On chainId: ", block.chainid);
        console.log("owner : ", msg.sender);

        if(block.chainid == 31337){
            vm.startBroadcast();
            VRFCoordinatorV2Mock(vrfCoordinator).fundSubscription(subId, FUND_AMOUNT);
            vm.stopBroadcast();
        }else{
            vm.startBroadcast();
            LinkToken(link).transferAndCall(vrfCoordinator, uint256(FUND_AMOUNT), abi.encode(link));
            vm.stopBroadcast();
        }
    }
    function run() external {
        return fundSubscriptionConfig();
    }
}

contract AddConsumer is Script{

    function addConsumer(address raffle, address vrfCoordinator, uint64 subId) public {
        console.log("Adding Consumer :", raffle);
        console.log("Using vrfCoordinator :", vrfCoordinator);
        console.log("On chainId", block.chainid);

        vm.startBroadcast();
        VRFCoordinatorV2Mock(vrfCoordinator).addConsumer(subId, raffle);
        vm.stopBroadcast();
    }

    function addConsumerConfig(address raffle) public {
        HelperConfig helperConfig = new HelperConfig();
        ( address vrfCoordinator,
        ,
        ,
        uint64 subId,
        ,
        ,
        ) = helperConfig.currentRaffleConfig();
        return addConsumer(raffle, vrfCoordinator, subId);
    }
    function run() external {
        address raffle = DevOpsTools.get_most_recent_deployment("Raffle", block.chainid);
        return addConsumerConfig(raffle);
    }
}`

Thanks for reply @Neftyr. Please help

**Reply by Parth-Sharma-tonchi** [reply]
Here I successfully run CreateSubscription script. But while running FundSubscription, it throws an error.

**Reply by Niferu** [reply]
@Parth-Sharma-tonchi 

Have you updated your subscription Id in helper accordingly?

**Reply by Parth-Sharma-tonchi** [reply]
Yes, I updated the subscription id in helperConfig. I run CreateSubscription and on UI my subscriptionId shows 3992.

HelperConfig.s.sol:

`//SPDX-License_Identifier: MIT

pragma solidity ^0.8.18;

import {VRFCoordinatorV2Mock} from "lib/chainlink-brownie-contracts/contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol";
import {Script} from "forge-std/Script.sol";
import {LinkToken} from "../test/mocks/LinkToken.sol";

contract HelperConfig is Script{

    RaffleConfig public currentRaffleConfig;

    struct RaffleConfig{
        address vrfCoordinatorV2;
        uint256 entranceFee;
        uint256 interval;
        uint64 subscriptionId;
        bytes32 gasLane;
        uint32 callbackGasLimit;
        address link;
    }

    constructor(){
        if(block.chainid == 11155111){
            currentRaffleConfig = getSepoliaRaffleConfig();
        }else{
            currentRaffleConfig = getLocalhostRaffleConfig();
        }
    }

    function getSepoliaRaffleConfig() public pure returns(RaffleConfig memory){
        return RaffleConfig({vrfCoordinatorV2:0x8103B0A8A00be2DDC778e6e7eaa21791Cd364625, entranceFee:0.01 ether, interval:30, **_subscriptionId:3992_**, gasLane:0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c, callbackGasLimit:500000, link:0x779877A7B0D9E8603169DdbD7836e478b4624789});
    }


    function getLocalhostRaffleConfig() public returns(RaffleConfig memory){
        if(currentRaffleConfig.vrfCoordinatorV2 != address(0)){
            return currentRaffleConfig;
        }
        uint96 baseFee = 0.25 ether;
        uint96 gasPriceLink = 1e9;

        vm.startBroadcast();
        VRFCoordinatorV2Mock vrfCoordinator = new VRFCoordinatorV2Mock(baseFee, gasPriceLink);

        LinkToken link = new LinkToken();
        
        vm.stopBroadcast();

        return RaffleConfig({vrfCoordinatorV2:address(vrfCoordinator), entranceFee:0.01 ether, interval:30,subscriptionId:1, gasLane:0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c, callbackGasLimit:500000,link: address(link)
        });
    }
}`

**Reply by Niferu** [reply]
Ok, so do you have enough LINK tokens in your wallet?

**Comment by Parth-Sharma-tonchi** [comment]
Yes, I have 26 LINK tokens in my Metamask wallet.

**Reply by Niferu** [reply]
Ok please link your repository, I will try to figure out what's wrong. Also make sure that you are using correct subscriptionId (check your current subscriptions Id's on chainlink website)

**Reply by Parth-Sharma-tonchi** [reply]
Thanks for your reply, sir. I got the answer below.

**Comment by alfheimrShiven** [comment, marked as answer]
Hey @Parth-Sharma-tonchi you're getting an EVM revert because the parameters being sent to  `LinkToken(link).transferAndCall(...)` function are incorrect. You're abi encoding `link` instead of `subId`.

Update your `transferAndCall()` call with the following:
```solidity 
LinkToken(link).transferAndCall(
                vrfCoordinator,
                uint256(FUND_AMOUNT),
                abi.encode(subId)
            );
```

This should work. Do let us know 🙃.

**Reply by Parth-Sharma-tonchi** [reply]
Thanks for your reply, Sir. It is helpful.

**Reply by alfheimrShiven** [reply]
Glad it helped @Parth-Sharma-tonchi and pls don't call me sir 😅. Shiven works 🙌🏾

**Comment by theirrationalone** [comment]
You're passing link instead of subId... 

```solidity
LinkToken(link).transferAndCall(vrfCoordinator,uint256(FUND_AMOUNT), abi.encode(link ❌));
```

Try this:

```solidity
LinkToken(link).transferAndCall(vrfCoordinator,uint256(FUND_AMOUNT), abi.encode(subId ✅));
```

**Reply by Parth-Sharma-tonchi** [reply]
Thanks for helping me, sir.
